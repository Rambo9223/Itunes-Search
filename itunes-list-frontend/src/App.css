.App {
  text-align: center;
  background: rgb(34,193,195);
  background: linear-gradient(0deg, rgb(16, 145, 147) 0%, rgb(124, 34, 198) 100%);
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-header {
  width:40%;
}

.App-body{
  width: 80%;
}

.title{
  color: black;
}

.add-container{
  padding: 20px 0px;
  position: relative;
  top: 20px;
}

#showlist{
  position: relative;
  top: 30px;
}

input[type=text], textarea {
  width: 100%; /* Full width */
  padding: 12px; /* Some padding */ 
  border: 1px solid #ccc; /* Gray border */
  border-radius: 4px; /* Rounded borders */
  box-sizing: border-box; /* Make sure that padding and width stays in place */
  margin-top: 6px; /* Add a top margin */
  margin-bottom: 16px; /* Bottom margin */
  resize: vertical /* Allow the user to vertically resize the textarea (not horizontally) */
}

.new-item-form{
  border-radius: 5px;
  background-color: #f2f2f2;
  padding: 20px;
  color: black;
}

.check-group{
  display: inline-flex;
  flex-direction: column;
  align-items:flex-start;
  justify-content: space-between;
  padding: 0 20px 0 0 ;
}
.check-container{
  position: relative;
  left: 25px;
}
.check-all{
  position: relative;
  bottom: 20px;
}
.favourite-container , .result-container{
  font-size: medium;
}

.nav-list{
  list-style: none;
  display: flex;
  flex-direction: row;
  align-items: center;
  text-decoration: none;
}

.nav-list>li{
  padding: 0 30px;
  color:inherit;
  text-decoration: none;
}

.nav-list>li:hover{
  color: rgb(65, 53, 242);
  text-decoration: underline;
  font-size: 28px;
}

