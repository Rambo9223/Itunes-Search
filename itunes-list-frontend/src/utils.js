import axios from "axios";
/*Base function to test if the api returns the correct data using the 
supplied base url */

export const BASE_URL = "https://itunes.apple.com/search?limit=1&term=all+the+small+things&media=music";

export const fetchData = async () => {
  try {
    return (await axios.get(`${BASE_URL}`));// if success return object
  } catch (e) {
    return [];// if unsucessful return empty array
  }
};