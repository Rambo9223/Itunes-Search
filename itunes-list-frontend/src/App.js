import React from "react";
import * as Icon from "react-bootstrap-icons"
import './App.css';
import "bootstrap/dist/css/bootstrap.min.css";
import GetFavourites from './Components/GetFavourites';
import NewSearch from './Components/NewSearch';
import { Route,Routes,Link } from "react-router-dom";
/* Marker Please note!!
as of today 29.05.23 the link in the course material 
https://performance-partners.apple.com/resources/documentation/itunes-store
returns a page with no content, if a student hadn't made their app yet 
there would be no instructions with how to set up search criteria to use the 
API.

Also not as per email with mentor I have not deployed this app.*/

/* Add a nav bar, that contains links to the pages 
for search and favourites, look in to how helmut works etc
 */
function App() {
  window.addEventListener("beforeunload", async function(event) {
  event.preventDefault();
  const response = await fetch('/api/last')
    if (!response.ok) {
    throw new Error('Data coud not be removed!')
    } else {
    this.alert("data removed");
    }
  });
  return (
    <>
    <div className="App">
      <h1 className="mb-3">My Itunes App <Icon.Apple/></h1>
      <nav className="nav-bar">
          <ul className="nav-list">
            <li><Link style={{textDecoration:"none",color:"white"}} className="links" to="/search" data-testid="Search">Search</Link></li>
            <li><Link style={{textDecoration:"none",color:"white"}} className="links" to="/favourites">My Favourites</Link></li>
          </ul>
        </nav>
        
    <Routes>
        <Route path="/search" element={<NewSearch/>}></Route>
        <Route path="/favourites" element={<GetFavourites/>}></Route>
      </Routes>
      </div>
      </> 
  );
}

export default App;
