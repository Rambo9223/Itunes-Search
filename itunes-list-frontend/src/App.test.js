import { render } from '@testing-library/react';
import App from './App';
import { BrowserRouter } from 'react-router-dom';
import React from 'react';
import mockAxios from "jest-mock-axios";
import { fetchData,BASE_URL } from './utils';


// test 1 App renders correctly and matches snapshot
function MockApp(){
  return (<>
  <BrowserRouter>
  <App/>
  </BrowserRouter>
  </>)
}

test("renders correctly",() => {
  const app = render(<MockApp/>);
  expect(app).toMatchSnapshot();
});

// Test 2 Unit test to check API return using mockAxios

describe("fetch API item", () => {
  afterEach(()=>{
    mockAxios.reset();
  });


  describe("when API call is successful", () => {
    it("should return the song object", async() => {
      // given
      const expected = [{
        "wrapperType": "track",
        "kind": "song",
        "artistId": 116851,
        "collectionId": 1440839912,
        "trackId": 1440840510,
        "artistName": "blink-182",
        "collectionName": "Enema of the State",
        "trackName": "All the Small Things",
        "collectionCensoredName": "Enema of the State",
        "trackCensoredName": "All the Small Things",
        "artistViewUrl": "https://music.apple.com/us/artist/blink-182/116851?uo=4",
        "collectionViewUrl": "https://music.apple.com/us/album/all-the-small-things/1440839912?i=1440840510&uo=4",
        "trackViewUrl": "https://music.apple.com/us/album/all-the-small-things/1440839912?i=1440840510&uo=4",
        "previewUrl": "https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview112/v4/48/18/c0/4818c0c9-4a95-eb7b-8632-1db6dd9b22dc/mzaf_3609150729841940164.plus.aac.p.m4a",
        "artworkUrl30": "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/c2/f2/78/c2f2783f-8e65-0b2b-80ac-b52c4cfece2d/16UMGIM35657.rgb.jpg/30x30bb.jpg",
        "artworkUrl60": "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/c2/f2/78/c2f2783f-8e65-0b2b-80ac-b52c4cfece2d/16UMGIM35657.rgb.jpg/60x60bb.jpg",
        "artworkUrl100": "https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/c2/f2/78/c2f2783f-8e65-0b2b-80ac-b52c4cfece2d/16UMGIM35657.rgb.jpg/100x100bb.jpg",
        "collectionPrice": 8.99,
        "trackPrice": 1.29,
        "releaseDate": "1999-01-01T12:00:00Z",
        "collectionExplicitness": "explicit",
        "trackExplicitness": "notExplicit",
        "discCount": 1,
        "discNumber": 1,
        "trackCount": 12,
        "trackNumber": 8,
        "trackTimeMillis": 167029,
        "country": "USA",
        "currency": "USD",
        "primaryGenreName": "Rock",
        "isStreamable": true
    }];
      mockAxios.get.mockResolvedValueOnce(expected);

      // result is the returned object from the utils.js function fetchData()
      const result = await fetchData();

      // then
      expect(mockAxios.get).toHaveBeenCalledWith(`${BASE_URL}`);
      expect(result).toEqual(expected);
    });
  });

  describe("when API call fails", () => {
    it("should return empty array", async () => {
      // given
      const message = "Network Error";
      mockAxios.get.mockRejectedValueOnce(new Error(message));

      // when
      const result = await fetchData();

      // then
      expect(mockAxios.get).toHaveBeenCalledWith(`${BASE_URL}`);
      expect(result).toEqual([]);
    });
  });
});
