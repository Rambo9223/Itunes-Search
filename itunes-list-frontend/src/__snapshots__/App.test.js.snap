// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <h1
          class="mb-3"
        >
          My Itunes App 
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.182.008C11.148-.03 9.923.023 8.857 1.18c-1.066 1.156-.902 2.482-.878 2.516.024.034 1.52.087 2.475-1.258.955-1.345.762-2.391.728-2.43Zm3.314 11.733c-.048-.096-2.325-1.234-2.113-3.422.212-2.189 1.675-2.789 1.698-2.854.023-.065-.597-.79-1.254-1.157a3.692 3.692 0 0 0-1.563-.434c-.108-.003-.483-.095-1.254.116-.508.139-1.653.589-1.968.607-.316.018-1.256-.522-2.267-.665-.647-.125-1.333.131-1.824.328-.49.196-1.422.754-2.074 2.237-.652 1.482-.311 3.83-.067 4.56.244.729.625 1.924 1.273 2.796.576.984 1.34 1.667 1.659 1.899.319.232 1.219.386 1.843.067.502-.308 1.408-.485 1.766-.472.357.013 1.061.154 1.782.539.571.197 1.111.115 1.652-.105.541-.221 1.324-1.059 2.238-2.758.347-.79.505-1.217.473-1.282Z"
            />
            <path
              d="M11.182.008C11.148-.03 9.923.023 8.857 1.18c-1.066 1.156-.902 2.482-.878 2.516.024.034 1.52.087 2.475-1.258.955-1.345.762-2.391.728-2.43Zm3.314 11.733c-.048-.096-2.325-1.234-2.113-3.422.212-2.189 1.675-2.789 1.698-2.854.023-.065-.597-.79-1.254-1.157a3.692 3.692 0 0 0-1.563-.434c-.108-.003-.483-.095-1.254.116-.508.139-1.653.589-1.968.607-.316.018-1.256-.522-2.267-.665-.647-.125-1.333.131-1.824.328-.49.196-1.422.754-2.074 2.237-.652 1.482-.311 3.83-.067 4.56.244.729.625 1.924 1.273 2.796.576.984 1.34 1.667 1.659 1.899.319.232 1.219.386 1.843.067.502-.308 1.408-.485 1.766-.472.357.013 1.061.154 1.782.539.571.197 1.111.115 1.652-.105.541-.221 1.324-1.059 2.238-2.758.347-.79.505-1.217.473-1.282Z"
            />
          </svg>
        </h1>
        <nav
          class="nav-bar"
        >
          <ul
            class="nav-list"
          >
            <li>
              <a
                class="links"
                data-testid="Search"
                href="/search"
                style="text-decoration: none; color: white;"
              >
                Search
              </a>
            </li>
            <li>
              <a
                class="links"
                href="/favourites"
                style="text-decoration: none; color: white;"
              >
                My Favourites
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <h1
        class="mb-3"
      >
        My Itunes App 
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.182.008C11.148-.03 9.923.023 8.857 1.18c-1.066 1.156-.902 2.482-.878 2.516.024.034 1.52.087 2.475-1.258.955-1.345.762-2.391.728-2.43Zm3.314 11.733c-.048-.096-2.325-1.234-2.113-3.422.212-2.189 1.675-2.789 1.698-2.854.023-.065-.597-.79-1.254-1.157a3.692 3.692 0 0 0-1.563-.434c-.108-.003-.483-.095-1.254.116-.508.139-1.653.589-1.968.607-.316.018-1.256-.522-2.267-.665-.647-.125-1.333.131-1.824.328-.49.196-1.422.754-2.074 2.237-.652 1.482-.311 3.83-.067 4.56.244.729.625 1.924 1.273 2.796.576.984 1.34 1.667 1.659 1.899.319.232 1.219.386 1.843.067.502-.308 1.408-.485 1.766-.472.357.013 1.061.154 1.782.539.571.197 1.111.115 1.652-.105.541-.221 1.324-1.059 2.238-2.758.347-.79.505-1.217.473-1.282Z"
          />
          <path
            d="M11.182.008C11.148-.03 9.923.023 8.857 1.18c-1.066 1.156-.902 2.482-.878 2.516.024.034 1.52.087 2.475-1.258.955-1.345.762-2.391.728-2.43Zm3.314 11.733c-.048-.096-2.325-1.234-2.113-3.422.212-2.189 1.675-2.789 1.698-2.854.023-.065-.597-.79-1.254-1.157a3.692 3.692 0 0 0-1.563-.434c-.108-.003-.483-.095-1.254.116-.508.139-1.653.589-1.968.607-.316.018-1.256-.522-2.267-.665-.647-.125-1.333.131-1.824.328-.49.196-1.422.754-2.074 2.237-.652 1.482-.311 3.83-.067 4.56.244.729.625 1.924 1.273 2.796.576.984 1.34 1.667 1.659 1.899.319.232 1.219.386 1.843.067.502-.308 1.408-.485 1.766-.472.357.013 1.061.154 1.782.539.571.197 1.111.115 1.652-.105.541-.221 1.324-1.059 2.238-2.758.347-.79.505-1.217.473-1.282Z"
          />
        </svg>
      </h1>
      <nav
        class="nav-bar"
      >
        <ul
          class="nav-list"
        >
          <li>
            <a
              class="links"
              data-testid="Search"
              href="/search"
              style="text-decoration: none; color: white;"
            >
              Search
            </a>
          </li>
          <li>
            <a
              class="links"
              href="/favourites"
              style="text-decoration: none; color: white;"
            >
              My Favourites
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
